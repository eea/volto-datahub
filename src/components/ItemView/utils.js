const OBSOLETE_KEYS = ['obsolete', 'superseded'];

export const isObsolete = (status) => {
  let obsolete = false;
  if (status) {
    if (Array.isArray(status)) {
      if (
        status.filter((stat) => OBSOLETE_KEYS.includes(stat?.key)).length > 0
      ) {
        obsolete = true;
      }
    } else {
      if (OBSOLETE_KEYS.includes(status?.key)) {
        obsolete = true;
      }
    }
  }
  return obsolete;
};

export const isInternalURL = (url) => {
  let internal = false;
  url.split('/').forEach((part) => {
    if (part.split('_')[6] === 'i') {
      internal = true;
    }
  });
  return internal;
};

export const isInternal = (item) => {
  let internal = false;
  const links = item?.link || [];
  links.forEach((link) => {
    if (isInternalURL(link?.url)) {
      internal = true;
    }
  });
  return internal;
};

export const groupBy = (obj, key) => {
  return obj.reduce((rv, item) => {
    const group = rv[item[key]] || [];
    group.push(item);
    rv[item[key]] = group;
    return rv;
  }, {});
};

export const SVGIcon = ({ name, size, color, className, title, onClick }) => {
  return (
    <svg
      xmlns={name.attributes && name.attributes.xmlns}
      viewBox={name.attributes && name.attributes.viewBox}
      style={{ height: size, width: 'auto', fill: color || 'currentColor' }}
      className={className ? `icon ${className}` : 'icon'}
      onClick={onClick}
      dangerouslySetInnerHTML={{
        __html: title ? `<title>${title}</title>${name.content}` : name.content,
      }}
    />
  );
};
