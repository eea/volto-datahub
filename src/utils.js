const OBSOLETE_KEYS = ['obsolete', 'superseded'];

export const isInternalURL = (url) => {
  let internal = false;
  url.split('/').forEach((part) => {
    if (part.split('_')[6] === 'i') {
      internal = true;
    }
  });
  return internal;
};

export const isObsolete = (status) => {
  let obsolete = false;
  if (status) {
    if (Array.isArray(status)) {
      if (
        status.filter((stat) => OBSOLETE_KEYS.includes(stat?.key)).length > 0
      ) {
        obsolete = true;
      }
    } else {
      if (OBSOLETE_KEYS.includes(status?.key)) {
        obsolete = true;
      }
    }
  }
  return obsolete;
};

export const isInternal = (item) => {
  let internal = false;
  const links = item?.link || [];
  links.forEach((link) => {
    if (isInternalURL(link?.url)) {
      internal = true;
    }
  });
  return internal;
};

export const groupBy = (obj, key) => {
  return obj.reduce((rv, item) => {
    const group = rv[item[key]] || [];
    group.push(item);
    rv[item[key]] = group;
    return rv;
  }, {});
};

export function parseDatasets(children) {
  const all_datasets = children.map((v, e) => {
    let date;
    if (v.resourceTemporalExtentDetails) {
      const tc = v.resourceTemporalExtentDetails[0];
      const { start, end } = tc || {};
      const tc_start = start?.date
        ? new Date(Date.parse(start.date)).getFullYear()
        : '';
      const tc_end = end?.date
        ? new Date(Date.parse(end.date)).getFullYear()
        : '';
      if (tc_start === tc_end) {
        date = tc_start;
      } else {
        date = tc_start + '-' + tc_end;
      }
    } else {
      date = '';
    }

    return Object.assign({ temporalDateRange: date, ...v }, e);
  });

  const datasets = all_datasets.filter((v) => {
    return !(isObsolete(v.cl_status) || isInternal(v));
  });

  const archivedOrRestrictedDatasets = all_datasets.filter((v) => {
    return isObsolete(v.cl_status) || isInternal(v);
  });

  const groupDatasets = (items) => {
    const datasetsByCoverage = groupBy(items, 'temporalDateRange');

    const sortedDatasets = Object.keys(datasetsByCoverage).sort(
      function (a, b) {
        a = Math.max(...a.split('-'));
        b = Math.max(...b.split('-'));
        return b - a;
      },
    );

    const groupedDatasets = sortedDatasets.map((v, i) => {
      return { id: i, date: v, children: datasetsByCoverage[v] };
    });

    return groupedDatasets;
  };

  const groupedByTemporalCoverage = groupDatasets(datasets);
  const groupedByArchivedOrRestricted = groupDatasets(
    archivedOrRestrictedDatasets,
  );

  return {
    all_datasets,
    datasets,
    archivedOrRestrictedDatasets,
    groupedByTemporalCoverage,
    groupedByArchivedOrRestricted,
  };
}
