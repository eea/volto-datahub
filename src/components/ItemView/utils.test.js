import { render, cleanup, fireEvent } from '@testing-library/react';
import { SVGIcon } from './utils';
import {
  isObsolete,
  isInternalURL,
  isInternal,
  groupBy,
} from '@eeacms/volto-datahub/utils';

describe('isObsolete', () => {
  it('should return true if status array contains obsolete keys', () => {
    const status = [{ key: 'obsolete' }, { key: 'active' }];
    expect(isObsolete(status)).toBe(true);
  });

  it('should return true if single status object key is obsolete', () => {
    const status = { key: 'superseded' };
    expect(isObsolete(status)).toBe(true);
  });

  it('should return false if status array does not contain obsolete keys', () => {
    const status = [{ key: 'active' }];
    expect(isObsolete(status)).toBe(false);
  });

  it('should return false if single status object key is not obsolete', () => {
    const status = { key: 'active' };
    expect(isObsolete(status)).toBe(false);
  });

  it('should return false if status is undefined', () => {
    const status = undefined;
    expect(isObsolete(status)).toBe(false);
  });
});

describe('isInternalURL', () => {
  it('should return true if a part of the URL contains "_i" as the seventh substring split by "_"', () => {
    const url = 'https://test.com/a/b/c/d/e/f/h_i_j_k_l_m_i_n/o/p';
    expect(isInternalURL(url)).toBe(true);
  });

  it('should return false if no part of the URL contains "_i" as the seventh substring split by "_"', () => {
    const url = 'https://test.com/a/b/c/d/e/f/h_i_j_k_l_m_n/o/p';
    expect(isInternalURL(url)).toBe(false);
  });
});

describe('isInternal', () => {
  it('should return true if any link in the item is internal', () => {
    const item = {
      link: [{ url: 'https://test.com/a/b/c/d/e/f/h_i_j_k_l_m_i_n/o/p' }],
    };
    expect(isInternal(item)).toBe(true);
  });

  it('should return false if link is undefined', () => {
    const item = {
      link: undefined,
    };
    expect(isInternal(item)).toBe(false);
  });

  it('should return false if no link in the item is internal', () => {
    const item = {
      link: [{ url: 'https://test.com/a/b/c/d/e/f/h_i_j_k_l_m_n/o/p' }],
    };
    expect(isInternal(item)).toBe(false);
  });
});

describe('groupBy', () => {
  it('should group array of objects by specified key', () => {
    const arr = [
      { id: 1, name: 'test1' },
      { id: 2, name: 'test2' },
      { id: 1, name: 'test3' },
    ];
    expect(groupBy(arr, 'id')).toEqual({
      '1': [
        { id: 1, name: 'test1' },
        { id: 1, name: 'test3' },
      ],
      '2': [{ id: 2, name: 'test2' }],
    });
  });
});

describe('SVGIcon', () => {
  afterEach(cleanup);

  it('should render SVGIcon correctly', () => {
    const { container } = render(
      <SVGIcon
        name={{
          attributes: {
            xmlns: 'http://www.w3.org/2000/svg',
            viewBox: '0 0 32 32',
          },
          content: '<path d="M2 9h28l-14 14L2 9z"></path>',
        }}
        className={'test'}
        title={'test title'}
        size="32px"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call onClick when clicked', () => {
    const onClick = jest.fn();
    const { container } = render(
      <SVGIcon
        name={{
          attributes: {
            xmlns: 'http://www.w3.org/2000/svg',
            viewBox: '0 0 32 32',
          },
          content: '<path d="M2 9h28l-14 14L2 9z"></path>',
        }}
        size="32px"
        onClick={onClick}
      />,
    );
    fireEvent.click(container.getElementsByClassName('icon')[0]);
    expect(onClick).toHaveBeenCalled();
  });
});
