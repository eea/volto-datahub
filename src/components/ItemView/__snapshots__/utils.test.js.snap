// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGIcon should render SVGIcon correctly 1`] = `
<svg
  class="icon test"
  style="height: 32px; width: auto; fill: currentColor;"
  viewBox="0 0 32 32"
  xmlns="http://www.w3.org/2000/svg"
>
  <title>
    test title
  </title>
  <path
    d="M2 9h28l-14 14L2 9z"
  />
</svg>
`;
